linode/manager
# M3-6039: Update FirewallRuleTable to support screen readers

## Description 📝

**What does this PR do?**
Refactors `FirewallRuleTable.tsx` to support screen readers by converting the underlying structure from a table into an unordered list arranged in a `Grid` to [improve compatibility with react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd/issues/2233). Added `aria-label`s and tab navigability where necessary to improve access to fields within a list item. 

Previously, the FirewallRuleTable was [inaccessible for users using screen readers and keyboard navigation](https://github.com/linode/manager/issues/7483#event-4432711610). The table did not always provide enough context for the user to know how to interact with drag & drop functionality or interact with the FirewallRuleAccessMenu buttons, and the fields in each table row were unknown to the user due to lack of focusability.

Detailed list of issue -> fixed behavior for each screen reader:

Both screen readers:
- No label name read for firewall rule row, just 'View details' ➡️ Screen reader reads firewall rule label when rule is focused
- ~No keyboard access to fields in firewall rule: Protocols, Port Range, Sources, Actions ➡️  Keyboard navigate to fields in a firewall rule when rule is focused~ - removed tab focus and made M3-6154; users can still view details about a rule by editing it in the meantime
- No table header focus - keyboard navigation jumped to first rule row with no context for the table ➡️ Keyboard navigate within list header group to read column headers

VoiceOver:
- No instructions about how to interact with drag & drop functionality read by screen reader ➡️ Screen reader reads drag & drop instructions when rule is focused

NVDA:
- Screen reader reads drag and drop instructions, but requires users to toggle focus mode before a firewall rule can be selected or any action buttons pressed. Screen reader instructions are confusing and misleading about dragging a row. ➡️ Screen reader still requires that user enter focus mode to activate a row... but an `aria-roledescription` added to the list item widget, mentioned in the [react-beautiful-dnd screen reader guide](https://github.com/atlassian/react-beautiful-dnd/blob/master/docs/guides/screen-reader.md), provides more context about how to interact with the draggable option.
 
## Preview 📷
<details>
<summary>
Screencaps with screen readers
</summary>
Prod - Mac/Chrome/VoiceOver:

https://user-images.githubusercontent.com/114685994/214673727-4a6b3e6a-124c-414e-8413-42fa125eadca.mov

This branch - Mac/Chrome/VoiceOver:

https://user-images.githubusercontent.com/114685994/214680631-9e482baa-3515-4bd0-a0f3-cb30bf4cb767.mov

Prod - Windows/Chrome/NVDA:

https://user-images.githubusercontent.com/114685994/214682408-555e0587-d508-44d4-a64d-a9b3ca18a3e0.mov

This branch - Windows/Chrome/NVDA:

https://user-images.githubusercontent.com/114685994/214682924-05875a6b-e180-41c3-92bd-a5f11d5012b0.mov
</details>

## How to test 🧪

**What are the steps to reproduce the issue or verify the changes?**
These changes need to be tested with multiple screen readers (VoiceOver, NVDA) and ideally on multiple browsers (Chrome, Firefox, Safari) to ensure the FirewallRuleTable functionality is accessible. 

VoiceOver
1. Turn on VoiceOver. Shortcut keys are `Command` + `F5` to toggle the screen reader. You can also find it in Mac settings under Accessibility.

NVDA:
1. Use a Windows machine installed with NVDA you have access to  _or_  set up a Windows emulator and install NVDA using our internal documentation on "Accessibility Testing with Virtual Machines and Screen Readers". (Ask me if you don't know where to find this.)
2. Turn on NVDA. Shortcut keys are `Control` + `Alt` + `N`.

With both screen readers:
1. Navigate to the Firewall Landing page.
2. Create a new firewall or use an existing one. Click on the firewall to view the Firewall Detail page.
3. Use the screen reader and keyboard navigation to navigate to either the inbound or outbound rule button. Create at least two new firewall rules with distinct labels (easiest way is to just set different presets, e.g. `SSH` and `HTTPS`). 
4. Navigate using the screen reader + keyboard to the appropriate firewall rules table (list).
5. Confirm that the above list of fixed behaviors are present using each screen reader. Firewall rules should be draggable and droppable, should be able to be edited, deleted, cloned, and reordered. Changes should persist as long as the changes are saved.
6. Confirm that the styling of the table and rules matches the styling in production.

**How do I run relevant unit or e2e tests?**
```yarn test FirewallTableRule```
